### PR TITLE
search: add tensor core to beam search space

### PR DESCRIPTION
**Pull request by flammit**
Screenshot 2024-01-29 at 2 29 34 PM

This depends on #3285 being merged first.

### CODE REVIEW

File: tinygrad/codegen/kernel.py [hunk 1]
```diff
@@ -390,16 +394,21 @@ def fix(needed, ax):
...
-        alias_pattern = [0]*(self.global_dims+(self.local_dims-len(tc.threads))) + [2]*(len(tc.threads)) + [0]*(self.shape_len-self.upcasted-self.first_reduce) + [1,1] + [3]*(self.upcasted-2)  # noqa: E501
-        self.alias_buffer(buf0, alias_pattern)
-        self.alias_buffer(buf1, alias_pattern)
+        # delay alias_buffer until all optimizations are completed
+        self.tensor_core_bufs = [buf0, buf1]
         return True
     return False
 
-  def apply_opt(self, opt:Opt):
+  def apply_opt(self, opt:Opt, append_opt:bool=True):
```
Comment (geohot) [inline tinygrad/codegen/kernel.py:402]: Hmm, so if I construct this with apply_tensor_cores it's never appended? I don't like this as an option, it should always append and be able to reconstruct.

Aka, the Linearizer state should always be able to be recreated with the ast + applied_opts

Comment (flammit) [reply tinygrad/codegen/kernel.py:402]: Whoops good catch - this wasn't intended.  Will push a fix shortly.

Comment (geohot) [reply tinygrad/codegen/kernel.py:402]: Can you add an explicit test for this reconstruction?

Comment (flammit) [reply tinygrad/codegen/kernel.py:382]: Added this test in `test_tensor_cores`

File: tinygrad/device.py [hunk 2]
```diff
@@ -249,12 +249,9 @@ def get_linearizer(self, ast:LazyOp) -> Linearizer:
...
     k.required_optimizations()
     if not NOOPT:
-      if not (used_tensor_cores:=k.apply_tensor_cores(getenv("TC", 1))): k.hand_coded_optimizations()
+      k.hand_coded_optimizations()
       if BEAM >= 1:
-        lins = [(("tc" if used_tensor_cores else "hc"), k)]
-        if used_tensor_cores:
-          lins.append(("hc", Linearizer(ast, self.compiler.linearizer_opts)))
-          lins[-1][1].hand_coded_optimizations()
+        lins = [("hc", k)]
```
Comment (geohot) [inline tinygrad/device.py:254]: I don't like that this doesn't try the hand coded and tensor cores separate anymore? Why did this have to change?

File: tinygrad/codegen/kernel.py [hunk 3]
```diff
@@ -470,6 +463,25 @@ def required_optimizations(self):
         self.apply_opt(Opt(OptOps.UPCAST, unit_stride_axes_mul_4[0], 4))
 
   def hand_coded_optimizations(self):
+    try: # check TC first and apply hand-coded opts if successful
```
Comment (geohot) [inline tinygrad/codegen/kernel.py:466]: Can this just work like how it used to work? I don't see why this has to be in hand coded.

File: tinygrad/features/search.py [hunk 4]
```diff
@@ -72,7 +73,7 @@ def bufs_from_lin(lin:Linearizer) -> List[Buffer]:
 
 # get dictionary of all possible actions
 def get_linearizer_actions(lin:Linearizer, include_0=True) -> Dict[int, Linearizer]:
-  acted_lins = {0:lin} if include_0 else {}
+  acted_lins, max_up, max_lcl = {0:lin} if include_0 else {}, getenv("BEAM_UP_MAX", 256), getenv("BEAM_LCL_MAX", 256)
```
[the remark this answers is not in the document]
Comment (flammit) [reply tinygrad/features/search.py:76]: Will remove from this PR and stage a separate PR for this later

**Comment by flammit** [comment]
@geohot is this closer to what you were expecting?

File: tinygrad/codegen/kernel.py [hunk 5]
```diff
@@ -119,13 +129,14 @@ def copy(self):
...
+    ret.info, ret.reduceop, ret.bufs, ret.earlybufs, ret.full_buf_index = \
+      self.info, self.reduceop, [x for x in self.bufs if not isinstance(x, LocalBuffer)], self.earlybufs, self.full_buf_index
+    ret.sts = self.sts[:len(ret.bufs)] # NOTE: must redo the local buffers with TC in beam
 
     # parameters for optimizations
-    ret.applied_opts, ret.group_for_reduces, ret.upcasted, ret.local_dims, ret.local_alias, ret.tensor_core, ret.dont_use_locals = \
-      self.applied_opts[:], self.group_for_reduces, self.upcasted, self.local_dims, self.local_alias.copy(), self.tensor_core, self.dont_use_locals
+    ret.applied_opts, ret.group_for_reduces, ret.upcasted, ret.local_dims, ret.dont_use_locals = \
+      self.applied_opts[:], self.group_for_reduces, self.upcasted, self.local_dims, self.dont_use_locals
+    ret.tensor_core, ret.tensor_core_opts, ret.local_alias = self.tensor_core, self.tensor_core_opts, {}
```
Comment (geohot) [inline tinygrad/codegen/kernel.py:139]: Ehh, why is this not `self.local_alias.copy()`?

Comment (flammit) [reply tinygrad/codegen/kernel.py:139]: the local_aliases depend on the full shape, so copying/then adding opts make them invalid.  Similar situation with the temp group_for_reduce buffer, which is why we drop out all LocalBuffer a couple lines above.

**Pull request merged**